File: python/indexer.py
from pathlib import Path
import json

DIRS = ['/Applications', '/Users/safwan']
EXCLUDED_PATHS = ['Library', 'opt', 'node_modules', '__pycache__', 'Icon?']
EXCLUDED_EXTS = ['.plist']


def path_black_list_conditions(file):
    return not (
        file.name.startswith('.') or
        any(file.match(d) for d in EXCLUDED_PATHS) or
        any(file.suffix == ext for ext in EXCLUDED_EXTS))


def indexer(dirs, index=None):
    if index is None:
        index = []

    for dir in dirs:
        for path in sorted(filter(path_black_list_conditions, Path(dir).iterdir())):
            if is_folder(path):
                indexer([path], index)

            index.append({
                'path': str(path.resolve()),
            })
        print(f'{dir} done.')
    return index


def is_folder(path):
    if path.suffix == '.app':
        return False
    try:
        list(path.iterdir())
    except (PermissionError, NotADirectoryError):
        return False
    return True


index = indexer(DIRS)
with open('index.json', 'w') as f:
    json.dump(index, f)

File: js/indexer.js
const fs = require('fs')
const path_module = require('path');
import *  as settings from './settings'

function indexer(dirs, index = null) {
    if (index === null) {
        index = [];
    }

    var subDirs;

    for (let dir of dirs) {
        subDirs = []
        for (let path of fs.readdirSync(dir).filter(pathBlackListConditions).sort()) {
            path = path_module.join(dir, path)
            if (is_folder(path)) indexer([path], index)
            index.push({
                path: fs.realpathSync(path),
            });
        }
    }
    return index;
}

function pathBlackListConditions(path) {
    return !(
        path_module.basename(path).startsWith('.') ||
        settings.EXCLUDED_PATHS.some(pattern => new RegExp(`${pattern}`).test(path)) ||
        settings.EXCLUDED_EXTS.some(ext => ext === path_module.extname(path))
    );
}

function is_folder(path) {
    if (settings.APP_EXTENTIONS.includes(path_module.extname(path))) {
        return false
    }
    try {
        fs.readdirSync(path)
    } catch (err) {
        return false;
    }
    return true;
}

fs.writeFileSync(process.argv[2], JSON.stringify(indexer(settings.DIRS)));
